File: 01_Python/practice/문자열연습1/22_isogram.py
# *******************************
# isogram : 중복글자 없는 단어
# isogram 여부 판단하기 (true/false)
# 

data = [
    "Dermatoglyphics", # -> True
    "programmer",      # -> False
    "Cocktail",        # -> False  대소문자 동일
    "isogram",         # -> True
]

def isIsogram(word):

    dict = {}

    for ch in word.lower():
        if ch.isalpha():
            dict[ch] = dict.get(ch, 0) + 1

    # return sum(dict.values()) == len(dict.values())

    for cnt in dict.values():
        if cnt > 1:
            word = "False"
            break
        else:
            word = "True"

    return word 

# def isIsogram(word):
#     word = word.lower()                     # word리스트 안에있는 문자열들을 전부 대소문자 통일
#     return len(word) == len(set(word)) 
    
print(list(map(isIsogram, data)))

# ↓ 결과 [True, False, False, True]

File: 03_Web/03_JavaScript/A_Basic/a35_문자열연습/22_isogram.js
/*******************************
 * isogram : 중복글자 없는 단어
 * isogram 여부 판단하기 (true/false)
 */

const input = [
    "Dermatoglyphics", // -> true
    "programmer",      // -> false
    "Cocktail",         // -> false  대소문자 동일
    "isogram",         // -> true

    // 다른 입력 데이터로 테스트 필요하면, 배열에 데이터 더 넣어보고 실행해도 됩니다
]


function isIsogram(str) {
    arr = str.toLowerCase().trim().split("")
    return !(str.toLowerCase().trim().split("")
        .some(ch => arr.splice(0, 1) && arr.includes(ch))) // 첫번쨰 값을 뺀 arr에 ch 가 있으면 바로 true 반환
}


input.forEach(e => console.log(isIsogram(e)));